dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+        Preconditions.checkArgument(result != null, "reader %s has not been analyzed.", reader);
+        return result;
+    }
+
+    @Override
+    public Writers getFinalWriters() {
+        final ReaderInfo result = readerMap.get(null);
+        Preconditions.checkState(result != null, "final state has not been analyzed.");
+        return result;
+    }
+
+    /**
+     * Lists all potential users of the result of an event.
+     * @param writer Event of interest to fetch information about.
+     * @throws IllegalArgumentException If {@code writer} was not inside the program when it was analyzed.
+     */
+    public Readers getReaders(RegWriter writer) {
+        Preconditions.checkNotNull(writer, "writer is null");
+        final Readers result = writerMap.get(writer);
+        Preconditions.checkArgument(result != null, "writer %s has not been analyzed.", writer);
+        return result;
+    }
+
+    /**
+     * Lists all potential users of uninitialized registers.
+     */
+    public Readers getInitialReaders() {
+        final Readers result = writerMap.get(null);
+        Preconditions.checkState(result != null, "initial state has not been analyzed.");
+        return result;
+    }
+
+    /**
+     * Analyzes an entire function definition.
+     * @param function Defined function to be analyzed.  Its parameters are marked as initialized.
+     */
+    public static BackwardsReachingDefinitionsAnalysis forFunction(Function function) {
+        final var analysis = new BackwardsReachingDefinitionsAnalysis();
+        final Set<Register> finalRegisters = new HashSet<>();
+        analysis.initialize(function, finalRegisters);
+        analysis.run(function, finalRegisters);
+        analysis.postProcess();
+        return analysis;
+    }
+
+    /**
+     * Analyzes an entire set of threads.
+     * <p>
+     * Optionally queries {@link ExecutionAnalysis} for pairs of writers appearing together in an execution.
+     * @param program Contains a set of threads to be analyzed.  Additionally-defined functions are ignored.
+     * @param analysisContext Collection of previous analyses to be used as dependencies.
+     * @param config User-defined settings for this analysis.
+     */
+    public static BackwardsReachingDefinitionsAnalysis fromConfig(Program program, Context analysisContext, Configuration config) {
+        final ExecutionAnalysis exec = analysisContext.get(ExecutionAnalysis.class);
+        final var analysis = new BackwardsReachingDefinitionsAnalysis();
+        final Set<Register> finalRegisters = finalRegisters(program);
+        for (Function function : program.isUnrolled() ? program.getThreads() :
+                Iterables.concat(program.getThreads(), program.getFunctions())) {
+            analysis.initialize(function, finalRegisters);
+            analysis.run(function, finalRegisters);
+        }
+        analysis.postProcess();
+        if (exec != null && program.isUnrolled()) {
```

Comment:
Can `exec` really be null?

Reply:
The analysis can be run during processing where `ExecutionAnalysis` is not available, but it can also be run afterwards. In the former case it should subsume `UseDefAnalysis` and in the latter case `Dependency`.

Reply:
Btw. do we get no must information at all without the `ExecutionAnalysis`?
Would a dominator tree analysis suffice to generate must information?